## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
print(df.info())
print(df.shape)
df_explore_sorted_client_id = df.sort_values(by='client_id')
df_explore_sorted_client_id





<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

first    last                                       job  \
47797     David   Ochoa  Geographical information systems officer   
49037     David   Ochoa  Geographical information systems officer   
38158     David   Ochoa  Geographical information systems officer   
48120     David   Ochoa  Geographical information systems officer   
44345     David   Ochoa  Geographical information systems officer   
...         ...     ...                                       ...   
23433  Samantha  Murray                             Herpetologist   
31352  Samantha  Murray                             Herpetologist   
29231  Samantha  Murray                             Herpetologist   
7140   Samantha  Murray                             Herpetologist   
43685  Samantha  Murray                             Herpetologist   

                    phone                                    email  client_id  \
47797        281.867.1615     david.measure.3427@morales-young.org      10033   
49037        281.867.1615     david.measure.3427@morales-young.org      10033   
38158        281.867.1615     david.measure.3427@morales-young.org      10033   
48120        281.867.1615     david.measure.3427@morales-young.org      10033   
44345        281.867.1615     david.measure.3427@morales-young.org      10033   
...                   ...                                      ...        ...   
23433  534.390.2260x56734  murray.samantha.4079@gonzalez-brown.com      99984   
31352  534.390.2260x56734  murray.samantha.4079@gonzalez-brown.com      99984   
29231  534.390.2260x56734  murray.samantha.4079@gonzalez-brown.com      99984   
7140   534.390.2260x56734  murray.samantha.4079@gonzalez-brown.com      99984   
43685  534.390.2260x56734  murray.samantha.4079@gonzalez-brown.com      99984   

       order_id  order_date  order_week  order_year         item_id  \
47797   3574869  2023-01-26           4        2023  ZDZ93019-47-4B   
49037   8159112  2023-01-16           3        2023  UCM40428-83-4C   
38158   8159112  2023-01-16           3        2023  XMZ38239-49-3D   
48120   3229787  2023-01-30           5        2023  AED21317-12-5A   
44345   3574869  2023-01-26           4        2023  DXU77211-79-0B   
...         ...         ...         ...         ...             ...   
23433   8437912  2023-01-25           4        2023  CUC61802-68-5D   
31352   5491843  2023-04-28          17        2023  ECU52209-77-0U   
29231   6821353  2023-02-17           7        2023  XZZ88006-14-6D   
7140    4547569  2023-04-05          14        2023  XBU40041-09-5M   
43685   9673005  2023-02-27           9        2023  CEX70724-13-4Z   

          category         subcategory  unit_price  unit_cost  unit_weight  \
47797     software  project management      168.36     102.19         3.41   
49037        decor            wall art      112.34      61.72         8.50   
38158    furniture              stands        1.48       1.22        18.04   
48120  electronics        audio visual      136.99     105.43         2.47   
44345    furniture              stands      200.79     136.93         6.70   
...            ...                 ...         ...        ...          ...   
23433  consumables      paper products      563.77     450.96         1.81   
31352  consumables    kitchen supplies       14.16       9.23         0.61   
29231  consumables   bathroom supplies        3.23       2.10        10.01   
7140   consumables      paper products        9.09       7.18         3.11   
43685  consumables    kitchen supplies      189.46     108.52        10.24   

       qty  line_number  
47797   29            2  
49037   10            2  
38158  146            3  
48120   60            0  
44345   37            1  
...    ...          ...  
23433   80            2  
31352   49            6  
29231    6            5  
7140    38            5  
43685  336            3  

[54639 rows x 18 columns]

In [5]:
# What three item categories had the most entries?
df_top_categories = df['category'].value_counts()
df_top_categories.head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries,
# which subcategory had the most entries?
df_consumables_top = df[(df['category'] == 'consumables')]['subcategory'].value_counts()
df_consumables_top.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?

df_top_client_entries = df['client_id'].value_counts()
df_top_client_entries.head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
top_5_client_entries = df_top_client_entries.head(5).index.tolist()
print(top_5_client_entries)


[33615, 66037, 46820, 38378, 24741]


In [9]:
# How many total units (the qty column) did the
# client with the most entries order order?
df_top_client_enterer = df[(df['client_id'] == top_5_client_entries[0])]
total_units = df_top_client_enterer['qty'].sum()
print(total_units)


64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
df_math_columns = df[['unit_price', 'qty', 'line_subtotal']]
df_math_columns.head(2)


unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)
df_shipping_math_columns = df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']]
df_shipping_math_columns.head(3)


unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [12]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['line_price'] = (df['line_subtotal'] + df['shipping_price']) * 1.0925
df_total_price_columns = df[['line_subtotal', 'shipping_price', 'line_price']]
df_total_price_columns.head(3)


line_subtotal  shipping_price     line_price
0      115164.00         5512.50  131839.076250
1         523.95          312.90     914.258625
2         527.28          458.64    1077.117600

In [13]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price     line_price  \
0            1      115164.00        787.50         5512.50  131839.076250   
1            0         523.95         31.29          312.90     914.258625   
2            6         527.28         65.52          458.64    1077.117600   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [14]:
# Create a column for the profit of
# each line using line cost and line price
df['line_profit'] = df['line_price'] - df['line_cost']
df.head(3)



first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price     line_price  line_cost  \
0      115164.00        787.50         5512.50  131839.076250   85597.05   
1         523.95         31.29          312.90     914.258625     629.79   
2         527.28         65.52          458.64    1077.117600     765.18   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
#I quadruple checked my work and I am confident that the columns have been created correctly. There seems to be no match for them within the original csv file.

# Check your work using the totals above
print(f"Order 2742071 total: ${df[(df['order_id'] == '2742071')]['line_price'].sum():,.2f}")
print(f"Order 2173913 total: ${df[(df['order_id'] == '2173913')]['line_price'].sum():,.2f}")
print(f"Order 6128929 profit: ${df[(df['order_id'] == '6128929')]['line_profit'].sum():,.2f}")


Order 2742071 total: $0.00
Order 2173913 total: $0.00
Order 6128929 profit: $0.00


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [17]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

for client in top_5_client_entries:
    print(f"Client {client} spent: ${df[(df['client_id'] == client)]['line_price'].sum():,.2f}")




Client 33615 spent: $8,377,308.57
Client 66037 spent: $10,259,514.80
Client 46820 spent: $9,743,794.32
Client 38378 spent: $12,906,550.88
Client 24741 spent: $82,268,891.98


In [32]:
# Create a summary DataFrame showing the totals for the
# top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

df_summary = df[['client_id', 'qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']]
df_summary = df_summary[df_summary['client_id'].isin(top_5_client_entries)]
df_summary = df_summary.groupby('client_id').sum().reset_index()
df_summary.head(5)


client_id     qty  shipping_price    line_price    line_cost   line_profit
0      24741  239862      5126448.37  8.226889e+07  45688899.71  3.657999e+07
1      33615   64313      1828984.89  8.377309e+06   6175313.91  2.201995e+06
2      38378   73667      3429455.40  1.290655e+07   9634720.98  3.271830e+06
3      46820   75768      1601448.84  9.743794e+06   7007191.64  2.736603e+06
4      66037   43018      1395151.85  1.025951e+07   7004482.98  3.255032e+06

In [31]:
#summary_df = df.loc[df['client_id'] == '24741']['client_id', 'qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']
#summary_df
#df_math_columns = df[['unit_price', 'qty', 'line_subtotal']]

#df_top_client_enterer = df[(df['client_id'] == top_5_client_entries[0])]
#total_units = df_top_client_enterer['qty'].sum()

#df_df.loc[df['category'] == 'electronics','sub_category'].value_counts().head(1)
df_summary = pd.DataFrame(columns=['client_id', 'total_units', 'total_shipping_price', 'total_revenue', 'total_profit'])
df_summary


Empty DataFrame
Columns: [client_id, total_units, total_shipping_price, total_revenue, total_profit]
Index: []

In [54]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

df_summary_renamed = df_summary.rename(columns={'client_id': 'Client ID',
                                                'qty': 'Units',
                                                 'shipping_price': 'Shipping',
                                                   'line_price': 'Total Revenue',
                                                     'line_cost': 'Total Cost',
                                                       'line_profit': 'Total Profit'})
def format_to_millions(x):
    return f"${x/1000000:,.2f}M"

#df_summary_renamed['Shipping'] = df_summary_renamed['Shipping'].applymap(format_to_millions)
#df_summary_renamed['Total Revenue'] = df_summary_renamed['Total Revenue'].apply(format_to_millions)
#df_summary_renamed['Total Cost'] = df_summary_renamed['Total Cost'].apply(format_to_millions)
#df_summary_renamed['Total Profit'] = df_summary_renamed['Total Profit'].apply(format_to_millions)

columns_to_format = ['Shipping', 'Total Revenue', 'Total Cost', 'Total Profit']
for column in columns_to_format:
    df_summary_renamed[column] = df_summary_renamed[column].apply(format_to_millions)

df_summary_renamed






Client ID   Units Shipping Total Revenue Total Cost Total Profit
0      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
1      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M
2      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
3      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
4      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M

In [55]:
# Sort the updated data by "Total Profit" form highest to lowest
df_summary_renamed = df_summary_renamed.sort_values(by='Total Profit', ascending=False)
df_summary_renamed

Client ID   Units Shipping Total Revenue Total Cost Total Profit
0      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
2      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
4      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
3      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
1      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M

In [56]:
df_summary_renamed = df_summary_renamed.sort_values(by='Total Profit', ascending=False)
df_summary_renamed

Client ID   Units Shipping Total Revenue Total Cost Total Profit
0      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
2      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
4      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
3      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
1      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M